File: src/domain/engine/engine.ts
import type {
  Board,
  Cell,
  GameEngine,
  GameRules,
  GameState,
  Move,
  Piece,
  PieceSize,
  Player,
  Position,
  Reserve,
} from '../types';
import { hashState } from './hash';

// ---------- internal helpers ----------

function otherPlayer(p: Player): Player {
  return p === 'P1' ? 'P2' : 'P1';
}

function emptyBoard(size: number): Board {
  return Array.from({ length: size }, () => Array.from({ length: size }, () => [] as Cell));
}

function makeReserve(rules: GameRules): Reserve {
  const r: Reserve = {};
  for (let i = 0; i < rules.pieceSizes.length; i++) {
    const ps = rules.pieceSizes[i];
    const cnt = rules.piecesPerSize[i];
    if (!ps || cnt === undefined) continue;
    r[ps.id] = cnt;
  }
  return r;
}

function cloneBoard(board: Board): Board {
  return board.map((row) => row.map((cell) => cell.slice()));
}

function cloneReserve(r: Reserve): Reserve {
  return { ...r };
}

function cloneState(state: GameState): GameState {
  return {
    rules: state.rules,
    board: cloneBoard(state.board),
    reserves: {
      P1: cloneReserve(state.reserves.P1),
      P2: cloneReserve(state.reserves.P2),
    },
    toMove: state.toMove,
    history: state.history.slice(),
    ply: state.ply,
    repetition: new Map(state.repetition),
    outcome: state.outcome,
  };
}

function rankMap(pieceSizes: readonly PieceSize[]): Map<string, number> {
  const m = new Map<string, number>();
  for (const ps of pieceSizes) m.set(ps.id, ps.rank);
  return m;
}

function topOf(cell: Cell): Piece | undefined {
  return cell.length > 0 ? cell[cell.length - 1] : undefined;
}

function inBounds(rules: GameRules, p: Position): boolean {
  return p.row >= 0 && p.row < rules.boardSize && p.col >= 0 && p.col < rules.boardSize;
}

/**
 * True iff a piece of `incomingRank` belonging to `incomingOwner` may land
 * on `targetCell` under the given rules. Empty cells are always allowed.
 */
function canCover(
  rules: GameRules,
  incomingOwner: Player,
  incomingRank: number,
  targetCell: Cell,
  ranks: Map<string, number>,
): boolean {
  const top = topOf(targetCell);
  if (!top) return true;
  const topRank = ranks.get(top.sizeId);
  if (topRank === undefined) return false;
  if (incomingRank < topRank) return false;
  if (incomingRank === topRank && !rules.allowSameSizeCover) return false;
  if (top.owner === incomingOwner && !rules.allowSelfCover) return false;
  return true;
}

// ---------- public engine ----------

function initialState(rules: GameRules): GameState {
  const state: GameState = {
    rules,
    board: emptyBoard(rules.boardSize),
    reserves: {
      P1: makeReserve(rules),
      P2: makeReserve(rules),
    },
    toMove: 'P1',
    history: [],
    ply: 0,
    repetition: new Map<string, number>(),
    outcome: null,
  };
  // Record the starting position once so repetition tracking is consistent.
  state.repetition.set(hashState(state), 1);
  return state;
}

function legalMoves(state: GameState): Move[] {
  if (state.outcome !== null) return [];
  const { rules, toMove, board } = state;
  const ranks = rankMap(rules.pieceSizes);
  const reserve = state.reserves[toMove];
  const moves: Move[] = [];

  // 1. Place from reserve.
  for (const ps of rules.pieceSizes) {
    const remaining = reserve[ps.id] ?? 0;
    if (remaining <= 0) continue;
    for (let r = 0; r < rules.boardSize; r++) {
      for (let c = 0; c < rules.boardSize; c++) {
        const cell = board[r]?.[c];
        if (!cell) continue;
        if (canCover(rules, toMove, ps.rank, cell, ranks)) {
          moves.push({
            kind: 'placeFromReserve',
            player: toMove,
            sizeId: ps.id,
            to: { row: r, col: c },
          });
        }
      }
    }
  }

  // 2. Move on board: pick up own top piece, drop on a different cell.
  for (let r = 0; r < rules.boardSize; r++) {
    for (let c = 0; c < rules.boardSize; c++) {
      const cell = board[r]?.[c];
      if (!cell) continue;
      const top = topOf(cell);
      if (top?.owner !== toMove) continue;
      const movingRank = ranks.get(top.sizeId);
      if (movingRank === undefined) continue;
      for (let r2 = 0; r2 < rules.boardSize; r2++) {
        for (let c2 = 0; c2 < rules.boardSize; c2++) {
          if (r === r2 && c === c2) continue;
          const target = board[r2]?.[c2];
          if (!target) continue;
          if (canCover(rules, toMove, movingRank, target, ranks)) {
            moves.push({
              kind: 'moveOnBoard',
              player: toMove,
              from: { row: r, col: c },
              to: { row: r2, col: c2 },
            });
          }
        }
      }
    }
  }

  return moves;
}

function applyMove(state: GameState, move: Move): GameState {
  if (state.outcome !== null) {
    throw new Error('applyMove: game has already ended');
  }
  if (move.player !== state.toMove) {
    throw new Error(`applyMove: ${move.player} cannot move; ${state.toMove} to move`);
  }
  const { rules } = state;
  if (!inBounds(rules, move.to)) {
    throw new Error('applyMove: target out of bounds');
  }
  if (move.kind === 'moveOnBoard' && !inBounds(rules, move.from)) {
    throw new Error('applyMove: source out of bounds');
  }

  const next = cloneState(state);
  const ranks = rankMap(rules.pieceSizes);

  if (move.kind === 'placeFromReserve') {
    const remaining = next.reserves[move.player][move.sizeId] ?? 0;
    if (remaining <= 0) {
      throw new Error(`applyMove: no ${move.sizeId} pieces in reserve`);
    }
    const targetCell = next.board[move.to.row]?.[move.to.col];
    if (!targetCell) throw new Error('applyMove: target cell missing');
    const ps = rules.pieceSizes.find((s) => s.id === move.sizeId);
    if (!ps) throw new Error(`applyMove: unknown sizeId ${move.sizeId}`);
    if (!canCover(rules, move.player, ps.rank, targetCell, ranks)) {
      throw new Error('applyMove: illegal cover');
    }
    targetCell.push({ owner: move.player, sizeId: move.sizeId });
    next.reserves[move.player][move.sizeId] = remaining - 1;
  } else {
    const fromCell = next.board[move.from.row]?.[move.from.col];
    const toCell = next.board[move.to.row]?.[move.to.col];
    if (!fromCell || !toCell) throw new Error('applyMove: cell missing');
    const top = topOf(fromCell);
    if (!top) throw new Error('applyMove: source cell empty');
    if (top.owner !== move.player) {
      throw new Error('applyMove: not own piece');
    }
    const movingRank = ranks.get(top.sizeId);
    if (movingRank === undefined) {
      throw new Error('applyMove: unknown size on board');
    }
    if (!canCover(rules, move.player, movingRank, toCell, ranks)) {
      throw new Error('applyMove: illegal cover (move on board)');
    }
    fromCell.pop();
    toCell.push(top);
  }

  next.history.push(move);
  next.ply += 1;
  next.toMove = otherPlayer(state.toMove);

  // Repetition tracking on the resulting position.
  const h = hashState(next);
  next.repetition.set(h, (next.repetition.get(h) ?? 0) + 1);

  // Update outcome eagerly so callers see terminal states.
  next.outcome = computeOutcome(next, h);

  return next;
}

function undo(state: GameState): GameState {
  if (state.history.length === 0) return state;
  const last = state.history[state.history.length - 1];
  if (!last) return state;

  const prev = cloneState(state);
  // Decrement repetition counter for the *current* (post-move) position.
  const currentHash = hashState(prev);
  const cnt = prev.repetition.get(currentHash) ?? 0;
  if (cnt <= 1) prev.repetition.delete(currentHash);
  else prev.repetition.set(currentHash, cnt - 1);

  // Reverse the move.
  if (last.kind === 'placeFromReserve') {
    const cell = prev.board[last.to.row]?.[last.to.col];
    if (!cell || cell.length === 0) {
      throw new Error('undo: target cell empty (corrupt state)');
    }
    const popped = cell.pop();
    if (popped?.owner !== last.player || popped.sizeId !== last.sizeId) {
      throw new Error('undo: top of stack does not match recorded move');
    }
    prev.reserves[last.player][last.sizeId] = (prev.reserves[last.player][last.sizeId] ?? 0) + 1;
  } else {
    const fromCell = prev.board[last.from.row]?.[last.from.col];
    const toCell = prev.board[last.to.row]?.[last.to.col];
    if (!fromCell || !toCell || toCell.length === 0) {
      throw new Error('undo: corrupt state for move-on-board');
    }
    const piece = toCell.pop();
    if (piece?.owner !== last.player) {
      throw new Error('undo: top piece is not the moved piece');
    }
    fromCell.push(piece);
  }

  prev.history.pop();
  prev.ply -= 1;
  prev.toMove = last.player;
  prev.outcome = null;
  return prev;
}

/**
 * Returns the winner if there is an N-in-a-row formed by top-of-stack
 * pieces (rows, columns, both diagonals). N is the board size.
 *
 * If both players somehow have a winning line on the same position
 * (impossible under normal play but defensive), the side that just moved
 * is preferred — but we simply return the first one we find since this
 * scenario does not arise with legal play.
 */
function isWin(state: GameState): Player | null {
  const N = state.rules.boardSize;
  const board = state.board;

  const tops: (Player | null)[][] = Array.from({ length: N }, (_, r) =>
    Array.from({ length: N }, (_, c) => {
      const cell = board[r]?.[c];
      const top = cell ? topOf(cell) : undefined;
      return top ? top.owner : null;
    }),
  );

  const checkLine = (cells: (Player | null)[]): Player | null => {
    if (cells.length === 0) return null;
    const first = cells[0];
    if (first === null || first === undefined) return null;
    for (const v of cells) if (v !== first) return null;
    return first;
  };

  // Rows & columns.
  for (let i = 0; i < N; i++) {
    const row = tops[i];
    if (row) {
      const w = checkLine(row);
      if (w) return w;
    }
    const col: (Player | null)[] = [];
    for (let j = 0; j < N; j++) col.push(tops[j]?.[i] ?? null);
    const wc = checkLine(col);
    if (wc) return wc;
  }

  // Diagonals.
  const diag1: (Player | null)[] = [];
  const diag2: (Player | null)[] = [];
  for (let i = 0; i < N; i++) {
    diag1.push(tops[i]?.[i] ?? null);
    diag2.push(tops[i]?.[N - 1 - i] ?? null);
  }
  const w1 = checkLine(diag1);
  if (w1) return w1;
  const w2 = checkLine(diag2);
  if (w2) return w2;

  return null;
}

/**
 * Compute the outcome for a state, given the (already computed) hash of
 * the current position. Win > maxPly > repetition > legal-moves draw.
 */
function computeOutcome(state: GameState, currentHash: string): Player | 'draw' | null {
  const winner = isWin(state);
  if (winner) return winner;
  if (state.ply >= state.rules.maxPly) return 'draw';
  const reps = state.repetition.get(currentHash) ?? 0;
  if (reps >= state.rules.drawByRepetition) return 'draw';
  // Cheap-to-detect "no legal moves" terminal — Gobblet rarely hits this
  // but the spec calls for it.
  // Use legalMoves directly; we are not in a hot loop yet.
  const stub: GameState = { ...state, outcome: null };
  if (legalMoves(stub).length === 0) return 'draw';
  return null;
}

function outcome(state: GameState): Player | 'draw' | null {
  if (state.outcome !== null) return state.outcome;
  return computeOutcome(state, hashState(state));
}

function isTerminal(state: GameState): boolean {
  return outcome(state) !== null;
}

function validateRules(rules: GameRules): { ok: boolean; errors: string[] } {
  const errors: string[] = [];

  if (!Number.isInteger(rules.boardSize) || rules.boardSize < 2) {
    errors.push('boardSize must be an integer >= 2');
  }
  if (!Array.isArray(rules.pieceSizes) || rules.pieceSizes.length === 0) {
    errors.push('pieceSizes must be a non-empty array');
  }
  if (!Array.isArray(rules.piecesPerSize)) {
    errors.push('piecesPerSize must be an array');
  }
  if (
    Array.isArray(rules.pieceSizes) &&
    Array.isArray(rules.piecesPerSize) &&
    rules.pieceSizes.length !== rules.piecesPerSize.length
  ) {
    errors.push('pieceSizes.length must equal piecesPerSize.length');
  }
  if (Array.isArray(rules.piecesPerSize)) {
    for (let i = 0; i < rules.piecesPerSize.length; i++) {
      const n = rules.piecesPerSize[i];
      if (!Number.isInteger(n) || n! <= 0) {
        errors.push(`piecesPerSize[${i}] must be a positive integer`);
      }
    }
  }
  if (Array.isArray(rules.pieceSizes)) {
    const seenIds = new Set<string>();
    const seenRanks = new Set<number>();
    for (let i = 0; i < rules.pieceSizes.length; i++) {
      const ps = rules.pieceSizes[i];
      if (!ps) continue;
      if (typeof ps.id !== 'string' || ps.id.length === 0) {
        errors.push(`pieceSizes[${i}].id must be a non-empty string`);
      } else if (seenIds.has(ps.id)) {
        errors.push(`pieceSizes[${i}].id duplicates "${ps.id}"`);
      } else {
        seenIds.add(ps.id);
      }
      if (!Number.isInteger(ps.rank) || ps.rank < 0) {
        errors.push(`pieceSizes[${i}].rank must be a non-negative integer`);
      } else if (seenRanks.has(ps.rank)) {
        errors.push(`pieceSizes[${i}].rank duplicates ${ps.rank}`);
      } else {
        seenRanks.add(ps.rank);
      }
    }
  }
  if (rules.winCondition.kind === 'lineOfN') {
    if (!Number.isInteger(rules.winCondition.n) || rules.winCondition.n <= 0) {
      errors.push('winCondition.n must be a positive integer');
    }
  }
  if (!Number.isInteger(rules.maxPly) || rules.maxPly <= 0) {
    errors.push('maxPly must be a positive integer');
  }
  if (!Number.isInteger(rules.drawByRepetition) || rules.drawByRepetition <= 0) {
    errors.push('drawByRepetition must be a positive integer');
  }
  return { ok: errors.length === 0, errors };
}

export const engine: GameEngine = {
  initialState,
  legalMoves,
  applyMove,
  undo,
  isTerminal,
  isWin,
  outcome,
  validateRules,
};

File: ai-training/src/dttt_train/engine.py
"""Pure-Python rule engine, mirror of the TypeScript engine.

Implements:

* ``legal_moves(state)``           - enumerate legal moves for ``state.to_move``
* ``apply_move(state, move)``      - return a new state after applying ``move``
* ``is_terminal(state)`` / ``outcome(state)``
* threefold-repetition + ``maxPly`` draw rules
* covering rules (``allow_same_size_cover`` / ``allow_self_cover``)
"""

from __future__ import annotations

import copy
from typing import Literal

from .rules import (
    Board,
    GameRules,
    GameState,
    MoveOnBoardMove,
    Piece,
    PlaceFromReserveMove,
    Player,
    initial_state,
)

Outcome = Player | Literal["draw"] | None


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _top(cell: list[Piece]) -> Piece | None:
    return cell[-1] if cell else None


def _can_cover(rules: GameRules, top: Piece | None, mover: Piece) -> bool:
    """Can ``mover`` legally land on ``top``?"""
    if top is None:
        return True
    mover_rank = rules.size_id_to_rank(mover.size_id)
    top_rank = rules.size_id_to_rank(top.size_id)
    if mover_rank > top_rank:
        # Larger always covers (subject to self-cover rule).
        if top.owner == mover.owner and not rules.allow_self_cover:
            return False
        return True
    if mover_rank == top_rank:
        if not rules.allow_same_size_cover:
            return False
        if top.owner == mover.owner and not rules.allow_self_cover:
            return False
        return True
    # Smaller never covers larger.
    return False


def _state_hash(state: GameState) -> str:
    """Canonical hash used for threefold-repetition detection.

    Encodes board stacks, reserves, and side-to-move; history is intentionally
    excluded.
    """
    parts: list[str] = []
    for row in state.board:
        for cell in row:
            parts.append("/".join(f"{p.owner.value}:{p.size_id}" for p in cell))
            parts.append("|")
        parts.append(";")
    parts.append("R1=")
    for ps in state.rules.piece_sizes:
        parts.append(f"{ps.id}{state.reserves[Player.P1].get(ps.id, 0)}")
    parts.append(",R2=")
    for ps in state.rules.piece_sizes:
        parts.append(f"{ps.id}{state.reserves[Player.P2].get(ps.id, 0)}")
    parts.append(f",T={state.to_move.value}")
    return "".join(parts)


def _check_winner(rules: GameRules, board: Board) -> Player | None:
    """Apply the current ``WinCondition`` (only ``lineOfN`` is implemented)."""
    n = rules.win_condition.n
    bs = rules.board_size

    def line_owner(cells: list[list[Piece]]) -> Player | None:
        if any(not c for c in cells):
            return None
        owners = {c[-1].owner for c in cells}
        return next(iter(owners)) if len(owners) == 1 else None

    # Rows
    for r in range(bs):
        for c0 in range(bs - n + 1):
            owner = line_owner([board[r][c0 + k] for k in range(n)])
            if owner is not None:
                return owner
    # Columns
    for c in range(bs):
        for r0 in range(bs - n + 1):
            owner = line_owner([board[r0 + k][c] for k in range(n)])
            if owner is not None:
                return owner
    # Diagonals \
    for r0 in range(bs - n + 1):
        for c0 in range(bs - n + 1):
            owner = line_owner([board[r0 + k][c0 + k] for k in range(n)])
            if owner is not None:
                return owner
    # Diagonals /
    for r0 in range(bs - n + 1):
        for c0 in range(n - 1, bs):
            owner = line_owner([board[r0 + k][c0 - k] for k in range(n)])
            if owner is not None:
                return owner
    return None


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def legal_moves(state: GameState) -> list[PlaceFromReserveMove | MoveOnBoardMove]:
    if state.outcome is not None:
        return []

    rules = state.rules
    bs = rules.board_size
    me = state.to_move
    moves: list[PlaceFromReserveMove | MoveOnBoardMove] = []

    # Place-from-reserve moves
    for ps in rules.piece_sizes:
        if state.reserves[me].get(ps.id, 0) <= 0:
            continue
        mover = Piece(owner=me, size_id=ps.id)
        for r in range(bs):
            for c in range(bs):
                top = _top(state.board[r][c])
                if _can_cover(rules, top, mover):
                    moves.append(
                        PlaceFromReserveMove(player=me, size_id=ps.id, to_row=r, to_col=c)
                    )

    # Move-on-board: only the player's own top piece may be lifted.
    for fr in range(bs):
        for fc in range(bs):
            top = _top(state.board[fr][fc])
            if top is None or top.owner != me:
                continue
            for tr in range(bs):
                for tc in range(bs):
                    if (fr, fc) == (tr, tc):
                        continue
                    target_top = _top(state.board[tr][tc])
                    if _can_cover(rules, target_top, top):
                        moves.append(
                            MoveOnBoardMove(
                                player=me,
                                from_row=fr,
                                from_col=fc,
                                to_row=tr,
                                to_col=tc,
                            )
                        )
    return moves


def apply_move(state: GameState, move: PlaceFromReserveMove | MoveOnBoardMove) -> GameState:
    """Return a new state with ``move`` applied. Pure (no mutation of input)."""
    if state.outcome is not None:
        raise ValueError("cannot apply a move to a terminal state")

    new_board: Board = [[list(cell) for cell in row] for row in state.board]
    new_reserves = {p: dict(r) for p, r in state.reserves.items()}

    if isinstance(move, PlaceFromReserveMove):
        if new_reserves[move.player].get(move.size_id, 0) <= 0:
            raise ValueError("no reserve piece of that size")
        new_reserves[move.player][move.size_id] -= 1
        new_board[move.to_row][move.to_col].append(
            Piece(owner=move.player, size_id=move.size_id)
        )
    else:
        src = new_board[move.from_row][move.from_col]
        if not src:
            raise ValueError("source cell is empty")
        piece = src.pop()
        if piece.owner != move.player:
            raise ValueError("can only move own piece")
        new_board[move.to_row][move.to_col].append(piece)

    new_history = list(state.history) + [move]
    new_ply = state.ply + 1
    new_to_move = state.to_move.opponent()

    new_state = GameState(
        rules=state.rules,
        board=new_board,
        reserves=new_reserves,
        to_move=new_to_move,
        history=new_history,
        ply=new_ply,
        repetition=copy.copy(state.repetition),
        outcome=None,
    )

    # Repetition / terminal detection
    h = _state_hash(new_state)
    new_state.repetition[h] = new_state.repetition.get(h, 0) + 1

    winner = _check_winner(state.rules, new_state.board)
    if winner is not None:
        new_state.outcome = winner
    elif new_state.repetition[h] >= state.rules.draw_by_repetition:
        new_state.outcome = "draw"
    elif new_state.ply >= state.rules.max_ply:
        new_state.outcome = "draw"
    elif not legal_moves(new_state):
        # Rare in Gobblet but defined for completeness.
        new_state.outcome = "draw"

    return new_state


def is_terminal(state: GameState) -> bool:
    return state.outcome is not None


def outcome(state: GameState) -> Outcome:
    return state.outcome


__all__ = [
    "initial_state",
    "legal_moves",
    "apply_move",
    "is_terminal",
    "outcome",
]
